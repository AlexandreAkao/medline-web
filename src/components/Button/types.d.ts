type ButtonSize = 'small' | 'medium' | 'large';

interface IButtonProps {
  size?: ButtonSize;
  children: React.ReactNode;
  onClick?: (event?: React.MouseEvent<HTMLButtonElement>) => void;
  primary?: boolean;
}
