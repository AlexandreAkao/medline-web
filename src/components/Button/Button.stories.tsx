import { ComponentStory, ComponentMeta } from '@storybook/react';

import Button from 'components/Button';

export default {
  title: 'Button',
  args: {
    size: 'medium',
    primary: true,
  },
  component: Button,
  argTypes: {
    size: {
      options: ['small', 'medium', 'large'],
      control: { type: 'radio' },
    },
  },
} as ComponentMeta<typeof Button>;

const Template: ComponentStory<typeof Button> = args => <Button {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  children: 'Primary',
  onClick: () => console.log('Clicked'),
};

export const Secondary = Template.bind({});
Secondary.args = {
  children: 'Secondary',
  onClick: () => console.log('Clicked'),
  primary: false,
};

export const Small = Template.bind({});
Small.args = {
  size: 'small',
  children: 'Small',
};

export const Large = Template.bind({});
Large.args = {
  size: 'large',
  children: 'Large',
};
