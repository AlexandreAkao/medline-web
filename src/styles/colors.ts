const colors = {
  primary: {
    normal: '#248F8D',
    light: '#279d9a',
    lighter: '#E9F4F4',
    darker: '#234342',
    darkest: '#257472',
  },
  white: {
    normal: '#fff',
    light: '#ccc',
    lighter: '#f0f0f0',
    lightest: '#fcfcfc',
  },
  black: {
    normal: '#000',
    soft: '#1d2129',
  },
  grey: {
    normal: '#808080',
    dark: '#3d3d3d',
  },
  blue: {
    normal: '#5041FF',
    light: '#0075FF',
  },
};

export default colors;
