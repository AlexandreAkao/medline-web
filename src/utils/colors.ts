export const getColorAlpha = (color: string, alpha = 1) => {
  const alphaLevel = Math.round(255 * alpha).toString(16);
  return color + alphaLevel;
};
