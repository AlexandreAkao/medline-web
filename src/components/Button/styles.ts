import styled, { css } from 'styled-components';

import colors from 'styles/colors';

const getButtonSize = (size: ButtonSize) => {
  const sizeMap = {
    small: '5px 20px',
    medium: '10px 40px',
    large: '10px 60px',
  };
  return sizeMap[size];
};

export const ButtonContainer = styled.button<Required<IButtonProps>>`
  border: 0;
  background-color: ${colors.primary};
  padding: ${({ size }) => getButtonSize(size)};
  border-radius: 20px;
  font-size: 1rem;
  transition: 0.2s ease all;
  color: ${colors.white.normal};
  font-weight: 700;
  cursor: pointer;
  display: inline-block;
  box-shadow: 0px 0px 15px 0px rgba(0, 0, 0, 0.2);

  :hover {
    transition: 0.2s ease all;
    background-color: ${colors.primaryLight};
  }

  ${({ primary }) =>
    !primary &&
    css`
      color: ${colors.black.normal};
      background-color: ${colors.white.normal};

      :hover {
        background-color: ${colors.white.lighter};
      }
    `}
`;
