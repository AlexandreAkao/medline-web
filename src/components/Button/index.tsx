import { ButtonContainer } from 'components/Button/styles';

function Button({ children, onClick, size = 'medium', primary = true }: IButtonProps) {
  return (
    <ButtonContainer onClick={onClick} size={size} primary={primary}>
      {children}
    </ButtonContainer>
  );
}

export default Button;
