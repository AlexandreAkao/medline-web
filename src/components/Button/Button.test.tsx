import { render, screen } from '@testing-library/react';
import { composeStories } from '@storybook/testing-react';
import userEvent from '@testing-library/user-event';
import { vi } from 'vitest';

import * as stories from 'components/Button/Button.stories';
import colors from 'styles/colors';
import Button from 'components/Button';

const { Primary, Secondary, Small, Large } = composeStories(stories);

describe('Button', () => {
  const renderComponent = (customProps: Partial<IButtonProps> = {}) => {
    const props: IButtonProps = {
      children: 'button example',
      ...customProps,
    };
    return render(<Button {...props} />);
  };

  it('should match with snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it('should render a primary Button', () => {
    render(<Primary />);
    const buttonElement = screen.getByText(/Primary/i);
    expect(buttonElement).not.toBeNull();
    expect(buttonElement).toHaveStyle({
      color: colors.white.normal,
      backgroundColor: colors.primary,
    });
  });

  it('should render a secondary Button', () => {
    render(<Secondary />);
    const buttonElement = screen.getByText(/Secondary/i);
    expect(buttonElement).not.toBeNull();
    expect(buttonElement).toHaveStyle({
      color: colors.black.normal,
      backgroundColor: colors.white.normal,
    });
  });

  it('should render a small Button', () => {
    render(<Small />);
    const buttonElement = screen.getByText(/Small/i);
    expect(buttonElement).not.toBeNull();
  });

  it('should render a large Button', () => {
    render(<Large />);
    const buttonElement = screen.getByText(/Large/i);
    expect(buttonElement).not.toBeNull();
  });

  it('should trigger function on click', async () => {
    const onClick = vi.fn();
    renderComponent({ onClick });
    const buttonElement = screen.getByText(/button example/i);
    await userEvent.click(buttonElement);
    expect(onClick).toBeCalled();
  });
});
