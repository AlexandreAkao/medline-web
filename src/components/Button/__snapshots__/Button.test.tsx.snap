// Vitest Snapshot v1

exports[`Button > should match with snapshot 1`] = `
<div>
  <button
    class="sc-bczRLJ eeLoKZ"
  >
    button example
  </button>
</div>
`;
